src/main/scala/org/apache/spark/sql/kinesis/HDFSMetadataCommitter.scala

```diff
@@ -165,4 +211,35 @@ class HDFSMetadataCommitter[T <: AnyRef : ClassTag](path: String,
     }
   }
 
+  /** Helper method to retry with exponential backoff  */
+  def withRetry[ T ](message: String, ignoreException: Boolean = true)(body: => T): T = {
+    var retryCount = 0
+    var result: Option[ T ] = None
+    var waitTimeInterval = retryIntervalMs
+    var lastError: Throwable = null
+
+    def isMaxRetryDone = retryCount >= numRetries
+
+    while (result.isEmpty && !isMaxRetryDone) {
+      if ( retryCount > 0 ) { // wait only if this is a retry
+        Thread.sleep(waitTimeInterval)
+        waitTimeInterval = scala.math.min(waitTimeInterval * 2, maxRetryIntervalMs)
+      }
+      try {
+        result = Some(body)
+      } catch {
+        case NonFatal(t) => lastError = t
+          if ( ignoreException ) {
+            logWarning(s"Error while $message [attempt = ${retryCount + 1}]", t)
+          } else {
+            throw new IllegalStateException(s"Error while $message", t)
+          }
+      }
+      retryCount += 1
+    }
+    result.getOrElse {
+      throw new IllegalStateException(s"Gave up after $retryCount retries while $message," +
```

Comment:
if it is a generic util method, should not we make this optional ( to throw exception). Should be responsibility of caller what needs to be done.

Reply:
Good point.  I don't think this `util` function can opt not to throw an exception.  This function is expected to return back the result of type `T`. It can retry a few times to do so and if its unable to get any result, it has to throw an exception. Its the caller responsibility to handle the exception and ignore it wishes to do so.  From the `util` POV, if I return "null", the caller has to be aware of this scenario and have to add checks for null in his code. Raising exception I think is better than silently ignoring the failures even after multiple retries.

Reply:
In that case, won't it be better to return Option[T] as output of this util method ? (with the argument method/body return Type as 'T' )
`def withRetry[ T ](message: String, ignoreException: Boolean = true)(body: => T):  Option[T] = {`

The reason why I am suggesting that is:

1. All Exceptions in scala are unchecked, even if we throw exception, caller method is not forced to handle the thrown exception , there will be no compile time error and he will get surprise at runtime (Similar to null scenario you mentioned)
Using Option as return type will solve that, caller method know upfront ,it might receive a None/null and will appropriately handle
2. Also, throwing exception is non-functional, it goes back to the stack trace and has to be caught/handled by the exact same caller even if he does not want to. Option will be functional, can be passed to some other caller without handling .

Reply:
Thanks for the feedback. I would like to go ahead with the current implementation. I want to throw exceptions since this is a retry helper function - either succeed or fail with an exception.